Implement basic stamina subsystem

Ryan089 opened the pull request:
## Summary

Implements a basic stamina subsystem. Features in this PR include:
- Depletion of stamina through movement
~~Depletion of stamina through interactions via the StaminaRequirement decorator.~~
- Regeneration of stamina over time.
- Ability to train stamina through use.
- A crappy UI display that should be removed or replaced ASAP.
- Basic unit tests

## Changes to Files

- Added StaminaController to **Human_Temporary** prefab.

- Added a weight field to **Item** script, and amended it to generate PickupHeavyInteractions rather than PickupInteractions. The weight of the item will determine the stamina cost of performing this interaction. Please note that this change can be reverted if desired, it is simply to demonstrate how a stamina requirement can be added to an existing interaction.

- Added StaminaBar to **Player Canvas** prefab to provide visual indicator of stamina.

- Amended **HumanoidController** script to require stamina when walking or running.

## Technical Notes

For interactions, stamina is server authoritative. For movement, stamina is "mock" server authoritative. The stamina subsystem should be refactored to be wholly server-authoritative when movement is transitioned across.

Stamina is shared between server and client via SyncVar. I have not currently restricted the throughput; however, it may be wise to do so as the game's network usage becomes more congested. Updating every 500-1000 milliseconds should be more than satisfactory.

I have included an allowable overdraw of 10% of the max stamina. Having no stamina should prevent you from starting an interaction, however, you *should* be able to continue the interaction until the overdraw limit is reached. The purpose of this is to prevent flipping rapidly between certain states (e.g. walking and running). Please note that this has not been completely implemented due to the "mock" server authoritative nature of the movement. As a result, when a player has no stamina but is still attempting to run, they will appear to move at a pace between running and walking.

Stamina regeneration changes during the game relative to the player's max stamina. This provides the appearance of constant stamina (both max and regeneration) but with stamina-depleting actions becoming easier as the game progresses. It is possible for stamina to regenerate faster than it is consumed by actions (like running) after stamina has been trained several times.

## Known issues

For some reason it doesn't seem like the player can drop items anymore. I don't think I caused that, but it is a possibility I guess.

## Review considerations / tips

- Change the weight of the BikeHorn or NukeCard in the inspector to test the stamina cost of picking up heavy items.

- Dramatically increase the TRAINING_MULTIPLIER (currently set at 1.05, or 5% increase in max stamina per training level) in Stamina.cs in order to more easily observe the "stamina level up" occurring. I have not exposed any API method to return the max stamina, only the current ratio.

## Fixes

Closes #949

Ryan089 committed:
Initial implementation of stamina use for walking / running.

Ryan089 committed:
Refactored StaminaController from SyncVar hook-driven event to regular polling. Stamina Bar now functional.

Ryan089 committed:
Created an interaction Requirement (StaminaRequirement) and tested with PickupHeavyInteraction. Weight field added to Item.cs

Ryan089 committed:
Allowed Stamina to be trained by using it.

Alainx277 commented on Assets/Scripts/EditorTests/HealthTests.cs:
```diff
@@ -0,0 +1,118 @@
+        public void CanCommenceInteractionOnlyWhenStaminaIsGreaterThatZero(float staminaToDeplete, bool expectedResult)
+        {
+
+            // ARRANGE
```
It's great that you're writing tests!

I think we can leave out the "step" comments, what do you think?

Ryan089 replied:
They're helpful for me as I develop them, but yeah, it doesn't need to be in the production code for tests this simple. I will remove them.

Alainx277 commented on Assets/Scripts/SS3D/Systems/Health/Indefatigable.cs:
```diff
@@ -0,0 +1,43 @@
+    /// sufficient stamina to perform any task where stamina is required.
+    /// </summary>
+    /// <inheritdoc cref="SS3D.Systems.Health.IStamina" />
+    public class Indefatigable : IStamina
```
I'm not sure I like there being an interface and an extra class for infinite stamina. Could we not omit the stamina class instead (leave null)?

Ryan089 replied:
Yep, I will remove the Indefatigable class and just manage with null. I would like to keep the IStamina interface though, it may come in helpful down the line.

Alainx277 commented on Assets/Scripts/SS3D/Systems/Health/StaminaController.cs:
```diff
@@ -0,0 +1,153 @@
+            }
+
+            // Currently movement is client-authoritative, so we need to subscribe to events on the client only.
+            if (IsClient && playerControllable.ControllingSoul.IsOwner)
+            {
+                SubscribeToEvents();
```
I don't think this will work when switching bodies. You could listen to `ControllingSoulChanged` on `playerControllable`

Ryan089 replied:
Thanks, I have reworked that and tested with MindSwap. It seems to work properly now.

Alainx277 commented on Assets/Scripts/SS3D/Systems/Health/StaminaRequirement.cs:
```diff
@@ -0,0 +1,83 @@
+        public override bool SatisfiesRequirement(InteractionEvent interactionEvent)
+        {
+            // Check to see if this source has a StaminaController component
+            IInteractionSource rootSource = InteractionSourceExtension.GetRootSource(interactionEvent.Source);
+            _controller = InteractionSourceExtension.GetComponent<StaminaController>(rootSource);
```
```suggestion
            _controller = interactionEvent.Source.GetRootSource(). GetComponent<StaminaController>();
```

Alainx277 commented on Assets/Scripts/SS3D/Systems/Health/StaminaRequirement.cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            if (_controller != null)
+            {
+                float staminaCost = 0f;
+                switch(exertion)
+                {
+                    case Exertion.None:
+                        break;
+                    case Exertion.Light:
+                        staminaCost = 1 * STAMINA_COST_PER_EXERTION_STEP;
+                        break;
+                    case Exertion.Moderate:
+                        staminaCost = 2 * STAMINA_COST_PER_EXERTION_STEP;
+                        break;
+                    case Exertion.Heavy:
+                        staminaCost = 3 * STAMINA_COST_PER_EXERTION_STEP;
+                        break;
+                    case Exertion.Extreme:
+                        staminaCost = 4 * STAMINA_COST_PER_EXERTION_STEP;
+                        break;
+                }
+
+                _controller.DepleteStaminaServer(staminaCost);
```
```suggestion
                _controller.DepleteStaminaServer((int)exertion * STAMINA_COST_PER_EXERTION_STEP);
```

Alainx277 commented on Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs:
```diff
@@ -176,7 +181,9 @@ private void OnDestroy()
         {
-            return new IInteraction[] { new PickupInteraction { Icon = _sprite } };
+            return new IInteraction[] { new PickupHeavyInteraction(_weight) { Icon = _sprite } };
+            //return new IInteraction[] { new PickupInteraction { Icon = _sprite } };
+
```
Please remove the commented code and empty line.

Alainx277 reviewed (changes requested):
Cool stuff!

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/IStamina.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        /// <param name="deltaTime"></param>
+        public void RechargeStamina(float deltaTime);
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/Indefatigable.cs:
```diff
@@ -0,0 +1,43 @@
+    /// <inheritdoc cref="SS3D.Systems.Health.IStamina" />
+    public class Indefatigable : IStamina
+    {
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/Stamina.cs:
```diff
@@ -0,0 +1,119 @@
+                // Scale up our current stamina. This is done to prevent visible stamina gap when training.
+                _current *= TRAINING_MULTIPLIER;
+            }
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaBarView.cs:
```diff
@@ -0,0 +1,51 @@
+            SetStamina(_controller ? _controller.Current : 0f);
+        }
+
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaController.cs:
```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        [SerializeField] private HumanoidController _player;
+
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaController.cs:
```diff
@@ -0,0 +1,153 @@
+                UI = FindObjectOfType<StaminaBarView>();
+                UI?.AssignViewToPlayer(this);
+            }
+
+            
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaHelper.cs:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    public enum StaminaType
+    {
+        Standard = 0,
+        Indefatigable = 1
+    }
```
Move to separate file, one file per declaration

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaRequirement.cs:
```diff
@@ -0,0 +1,83 @@
+        /// </summary>
+        private StaminaController _controller;
+
+        private const float STAMINA_COST_PER_EXERTION_STEP = 2.5f;
```
Consts should follow UpperCaseCamelCase pattern

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaRequirement.cs:
```diff
@@ -0,0 +1,83 @@
+
+namespace SS3D.Systems.Health
+{
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaRequirement.cs:
```diff
@@ -0,0 +1,83 @@
+
+            // Permit this interaction if the source is not subject to stamina, or if it has sufficient to interact.
+            return (_controller == null || _controller.CanCommenceInteraction);
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/StaminaRequirement.cs:
```diff
@@ -0,0 +1,83 @@
+
+    }
+
+    public enum Exertion : int
+    {
+        None = 0,
+        Light = 1,
+        Moderate = 2,
+        Heavy = 3,
+        Extreme = 4
+    }
```
Move to another file, one file per declaration

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/Interactions/PickupHeavyInteraction.cs:
```diff
@@ -0,0 +1,55 @@
+            {
+                exertion = Exertion.Extreme;
+            }
+
```
Remove empty space

joaoburatto commented on Assets/Scripts/SS3D/Systems/Storage/Interactions/PickupHeavyInteraction.cs:
```diff
@@ -0,0 +1,55 @@
+            set => (Interaction as PickupInteraction).Icon = value;
+        }
+    }
+
+
+
```
Remove empty space

TaylorNAlbarnaz commented:
Imo the system shouldn't remove stamina immediately through usage, it should work by slowing down the stamina regen and speeding up the stamina depletion, nothing else, for now it could be just by holding an heavy item

Ryan089 committed:
Removed stamina requirement decorator for interactions, reverted Item script.
Corrected MindSwap behaviour for stamina UI.
Corrected code style inconsistencies identified in review.
